src/openrct2-ui/windows/Research.cpp

```diff
@@ -154,117 +113,218 @@ static constexpr const StringId ResearchStageNames[] = {
     STR_RESEARCH_STAGE_UNKNOWN,
 };
 
-static void WindowResearchSetPage(WindowBase* w, int32_t page);
-static void WindowResearchSetPressedTab(WindowBase* w);
-static void WindowResearchDrawTabImages(DrawPixelInfo& dpi, WindowBase* w);
-
-WindowBase* WindowResearchOpen()
+class ResearchWindow final : public Window
 {
-    WindowBase* w;
-
-    w = WindowBringToFrontByClass(WindowClass::Research);
-    if (w == nullptr)
+    void OnOpen() override
     {
-        w = WindowCreateAutoPos(WW_FUNDING, WH_FUNDING, window_research_page_events[0], WindowClass::Research, WF_10);
-        w->widgets = window_research_page_widgets[0];
-        w->number = 0;
-        w->page = 0;
-        w->frame_no = 0;
-        w->disabled_widgets = 0;
+        SetPage(WINDOW_RESEARCH_PAGE_DEVELOPMENT);
         ResearchUpdateUncompletedTypes();
     }
 
-    w->page = 0;
-    w->Invalidate();
-    w->width = WW_DEVELOPMENT;
-    w->height = WH_DEVELOPMENT;
-    w->Invalidate();
+    void SetPage(int32_t p)
+    {
+        page = p;
+        frame_no = 0;
+        RemoveViewport();
 
-    w->widgets = window_research_page_widgets[0];
-    w->hold_down_widgets = 0;
-    w->event_handlers = window_research_page_events[0];
-    w->pressed_widgets = 0;
-    w->disabled_widgets = 0;
-    WindowInitScrollWidgets(*w);
+        hold_down_widgets = 0;
+        widgets = window_research_page_widgets[p];
+        disabled_widgets = 0;
+        pressed_widgets = 0;
 
-    return w;
-}
+        Invalidate();
+        if (p == WINDOW_RESEARCH_PAGE_DEVELOPMENT)
+        {
+            width = WW_DEVELOPMENT;
+            height = WH_DEVELOPMENT;
+        }
+        else
+        {
+            width = WW_FUNDING;
+            height = WH_FUNDING;
+        }
+        WindowEventResizeCall(this);
+        WindowEventInvalidateCall(this);
 
-#pragma region Development page
+        WindowInitScrollWidgets(*this);
+        Invalidate();
+    }
 
-/**
- *
- *  rct2: 0x006B6B38
- */
-static void WindowResearchDevelopmentMouseup(WindowBase* w, WidgetIndex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnUpdate() override
     {
-        case WIDX_CLOSE:
-            WindowClose(*w);
-            break;
-        case WIDX_TAB_1:
-        case WIDX_TAB_2:
-            WindowResearchSetPage(w, widgetIndex - WIDX_TAB_1);
-            break;
-        case WIDX_LAST_DEVELOPMENT_BUTTON:
-            News::OpenSubject(News::ItemType::Research, gResearchLastItem->rawValue);
-            break;
+        // Tab animation
+        if (++frame_no >= window_research_tab_animation_loops[page])
+            frame_no = 0;
+        InvalidateWidget(WIDX_TAB_1 + page);
```

Comment:
Maybe:
```suggestion
        InvalidateWidget(page == WINDOW_RESEARCH_PAGE_DEVELOPMENT ? WIDX_TAB_1 : WIDX_TAB_2);
```

But not sure if that is really better

Reply:
Maybe a clearer more verbose approach?
```suggestion
        switch (page)
        {
            case WINDOW_RESEARCH_PAGE_DEVELOPMENT:
            {
                InvalidateWidget(WIDX_TAB_1);
                break;
            }
            case WINDOW_RESEARCH_PAGE_FUNDING:
            {
                InvalidateWidget(WIDX_TAB_2);
                break;
            }
        }
```

Reply:
Hmm, I don't know about this one, maybe it is too verbose? Though technically less error prone